src/dxil/wrapper.rs

```diff
@@ -0,0 +1,515 @@
+        D3D12ShaderReflectionVariable::new(ptr)
+    }
+}
+
+#[derive(Debug)]
+pub struct D3D12ShaderReflectionType {
+    inner: ComPtr<ID3D12ShaderReflectionType>,
+}
+
+impl D3D12ShaderReflectionType {
+    fn new(inner: ComPtr<ID3D12ShaderReflectionType>) -> Self {
+        Self { inner }
+    }
+
+    pub fn get_base_class(&self) -> D3D12ShaderReflectionType {
+        let mut ptr: ComPtr<ID3D12ShaderReflectionType> = ComPtr::new();
+        unsafe { *ptr.as_mut_ptr() = self.inner.get_base_class() };
+        D3D12ShaderReflectionType::new(ptr)
+    }
+
+    pub fn get_desc(&self) -> Result<d3d12shader::D3D12_SHADER_TYPE_DESC, HassleError> {
+        let mut desc = d3d12shader::D3D12_SHADER_TYPE_DESC::default();
+        check_hr_wrapped!(unsafe { self.inner.get_desc(&mut desc as *mut _) }, desc)
+    }
+
+    pub fn get_inferface_by_index(&self, index: u32) -> D3D12ShaderReflectionType {
+        let mut ptr: ComPtr<ID3D12ShaderReflectionType> = ComPtr::new();
+        unsafe { *ptr.as_mut_ptr() = self.inner.get_interface_by_index(index) };
+        D3D12ShaderReflectionType::new(ptr)
+    }
+
+    pub fn get_member_type_by_index(&self, index: u32) -> D3D12ShaderReflectionType {
+        let mut ptr: ComPtr<ID3D12ShaderReflectionType> = ComPtr::new();
+        unsafe { *ptr.as_mut_ptr() = self.inner.get_member_type_by_index(index) };
+        D3D12ShaderReflectionType::new(ptr)
+    }
+
+    pub fn get_member_type_by_name(&self, name: &CStr) -> D3D12ShaderReflectionType {
+        let mut ptr: ComPtr<ID3D12ShaderReflectionType> = ComPtr::new();
+        unsafe {
+            *ptr.as_mut_ptr() = self
+                .inner
+                .get_member_type_by_name(name.as_ptr() as *const i8)
+        };
+        D3D12ShaderReflectionType::new(ptr)
+    }
+
+    pub fn get_member_type_name(&self, index: u32) -> Result<&str, HassleError> {
+        unsafe {
+            CStr::from_ptr(self.inner.get_member_type_name(index))
+                .to_str()
+                .map_err(HassleError::Utf8Error)
+        }
+    }
+
+    pub fn get_num_interfaces(&self) -> u32 {
+        unsafe { self.inner.get_num_interfaces() }
+    }
+
+    pub fn get_sub_type(&self) -> D3D12ShaderReflectionType {
+        let mut ptr: ComPtr<ID3D12ShaderReflectionType> = ComPtr::new();
+        unsafe { *ptr.as_mut_ptr() = self.inner.get_sub_type() };
+        D3D12ShaderReflectionType::new(ptr)
+    }
+
+    pub fn implements_interface(&self, base: &D3D12ShaderReflectionType) -> HRESULT {
+        unsafe {
+            self.inner
+                .implements_interface(*base.inner.clone().as_mut_ptr())
+        }
+    }
+
+    pub fn is_equal(&self, desc: &D3D12ShaderReflectionType) -> HRESULT {
+        unsafe { self.inner.is_equal(*desc.inner.clone().as_mut_ptr()) }
+    }
+
+    pub fn is_of_type(&self, desc: &D3D12ShaderReflectionType) -> HRESULT {
+        unsafe { self.inner.is_of_type(*desc.inner.clone().as_mut_ptr()) }
```

Comment:
simply calling .as_ptr() does not seem to work, however `base.inner.as_ptr::<ID3D12ShaderReflectionType>() as *mut _` does seem to work removing clone().

Reply:
Ah, that looks to be because of FFI taking these as `*mut`. Then `base.inner.as_ptr() as *mut _` should work as well, right?

Anyway, I think the right course of action is to turn these FFI functions into `*const` as they should not be modifying the argument.